app/src/main/java/de/grobox/transportr/trips/search/DirectionsFragment.kt

```diff
@@ -0,0 +1,274 @@
+        viewModel.isDeparture.observe(viewLifecycleOwner, Observer<Boolean> { this.onIsDepartureChanged(it) })
+        viewModel.isExpanded.observe(viewLifecycleOwner, Observer<Boolean> { this.onViaVisibleChanged(it) })
+        viewModel.calendar.observe(viewLifecycleOwner, Observer { this.onCalendarUpdated(it) })
+        viewModel.findGpsLocation.observe(viewLifecycleOwner, Observer { this.onFindGpsLocation(it) })
+        viewModel.isFavTrip.observe(viewLifecycleOwner, Observer { this.onFavStatusChanged(it) })
+        viewModel.products.observe(viewLifecycleOwner, Observer { this.onProductsChanged(it) })
+
+        favIcon.setOnClickListener { viewModel.toggleFavTrip() }
+
+        timeBackground.setOnClickListener {
+            if (viewModel.calendar.value == null) throw IllegalStateException()
+            val fragment = TimeDateFragment.newInstance(viewModel.calendar.value!!, viewModel.isDeparture.value!!)
+            fragment.setTimeDateListener(viewModel)
+            fragment.show(activity!!.supportFragmentManager, TimeDateFragment.TAG)
+        }
+        timeBackground.setOnLongClickListener {
+            viewModel.resetCalender()
+            true
+        }
+
+        productsIcon.setOnClickListener {
+            activity?.let { a ->
+                ProductDialogFragment().show(a.supportFragmentManager, ProductDialogFragment.TAG)
+            }
+        }
+        swapIcon.setOnClickListener { swapLocations() }
+        viaIcon.setOnClickListener { viewModel.toggleIsExpanded() }
+
+        TooltipCompat.setTooltipText(productsIcon, getString(R.string.action_choose_products))
+        TooltipCompat.setTooltipText(swapIcon, getString(R.string.action_switch_locations))
+        TooltipCompat.setTooltipText(viaIcon, getString(R.string.action_navigation_expand))
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putSerializable(DATE, viewModel.calendar.value)
+        outState.putBoolean(EXPANDED, viewModel.isExpanded.value ?: false)
+        outState.putBoolean(DEPARTURE, viewModel.isDeparture.value ?: true)
+    }
+
+    override fun onViewStateRestored(savedInstanceState: Bundle?) {
+        super.onViewStateRestored(savedInstanceState)
+        if (savedInstanceState != null && viewModel.trips.value == null) {
+            viewModel.setIsExpanded(savedInstanceState.getBoolean(EXPANDED, false))
+            viewModel.setIsDeparture(savedInstanceState.getBoolean(DEPARTURE, true))
+            viewModel.setFromLocation(fromLocation.getLocation())
+            viewModel.setViaLocation(viaLocation.getLocation())
+            viewModel.setToLocation(toLocation.getLocation())
+            viewModel.onTimeAndDateSet(savedInstanceState.getSerializable(DATE) as Calendar)
+        }
+        (activity!!.supportFragmentManager.findFragmentByTag(TimeDateFragment.TAG) as? TimeDateFragment)?.setTimeDateListener(viewModel)
```

Comment:
Not necessary now, but eventually we should get rid of listeners like this, they are flaky. Using the same ViewModel and LiveData is better.

Reply:
Opened #666